lib/Analysis/Allocation.cpp

```diff
@@ -374,35 +452,82 @@ class AllocationAnalysis {
     });
   }
 
+  void updateBufferRange(BufferT *buffer, IntervalT interval) {
+    if (bufferRange.contains(buffer))
+      interval = interval.merge(bufferRange[buffer]);
+    bufferRange[buffer] = interval;
+    LLVM_DEBUG(llvm::dbgs() << "    UPDATE BUFFER: " << buffer->id << " -> "
+                            << interval << "\n");
+  }
+
   /// Computes the liveness range of the allocated value.
   /// Each buffer is allocated only once.
-  void resolveExplicitBufferLiveness(
-      function_ref<Interval<size_t>(Value value)> getLiveness) {
+  void resolveExplicitBufferLiveness(LivenessF getLiveness) {
     for (auto valueBufferIter : allocation->valueBuffer) {
       auto value = valueBufferIter.first;
       auto *buffer = valueBufferIter.second;
-      bufferRange[buffer] = getLiveness(value);
+      auto ranges = getLiveness(value);
+      // TODO(SJW): bufferRange should support disjoint intervals
+      // This is only a problem for scf loops; cf branches handle
+      // this with aliases
+      if (!ranges.empty())
+        updateBufferRange(buffer, ranges.span());
     }
   }
 
-  /// Extends the liveness range by unionizing the liveness range of the aliased
-  /// values because each allocated buffer could be an alias of others, if block
-  /// arguments are involved.
-  void resolveAliasBufferLiveness(
-      function_ref<Interval<size_t>(Value value)> getLiveness) {
+  void dumpAliasRanges(Value value, const llvm::SetVector<BufferT *> &buffers,
+                       const LiveIntervals &ranges) const {
+    llvm::dbgs() << "    RESOLVE ALIAS: "
+                 << "\n\t";
+    printValue(value);
+    llvm::dbgs() << "\n\tRANGES:   ";
+    llvm::for_each(ranges, [cnt = 0](auto interval) mutable {
+      llvm::dbgs() << (cnt++ ? ", " : "") << interval;
+    });
+    llvm::dbgs() << "\n\tBUFFERS:  ";
+    llvm::for_each(buffers, [cnt = 0](auto *buf) mutable {
+      llvm::dbgs() << (cnt++ ? ", " : "") << buf->id;
+    });
+    llvm::dbgs() << "\n";
+  }
+
+  /// Following the alias lattice, an alias that only references a single
+  /// buffer, can simply update the live range of that buffer.
+  /// When an alias references multiple buffers as in the case of loop-
+  /// carried variables, create a new Buffer<Alias> for each disjoint range.
+  /// This new Buffer<Alias> will represent the live-range for all referenced
+  /// buffers.
+  /// Example: (numbers represent op liveness index)
+  ///  3    %b0 = convert_layout %g0                  -- Buffer #0
+  ///  4    %fr = for (.., %arg0 = %b0) {             -+ Alias #1
+  ///  5        %gn = load %pc                         |   Buffers #0,#1
+  ///  6        %bc = convert_layout %arg0            -+
+  ///  7        %v = add %bc, ...
+  ///  8        %bn = convert_layout %gn              -+ Buffer #1
+  ///  9        %pn = addptr %pc, %cst                 |   loop-carried
```

Comment:
I'm probably still confusing about the command. Do you want to say buffer `#1` is loop-carried or `%pn` is loop-carried?